File: mobile/ios/FUN/FUN/Views/Components/UnlockSheet.swift
/**
 * FUN App - Unlock Sheet
 * Bottom sheet with 4 unlock methods
 */

import SwiftUI

struct UnlockSheet: View {
    let episode: Episode
    let seriesId: String
    @Environment(\.dismiss) var dismiss
    @EnvironmentObject var authViewModel: AuthViewModel
    
    @State private var isUnlocking = false
    @State private var errorMessage: String?
    
    var body: some View {
        NavigationView {
            VStack(spacing: 0) {
                // Header
                VStack(spacing: 8) {
                    Text("Unlock Episode \(episode.episodeNum)")
                        .font(.title2.weight(.bold))
                        .foregroundColor(Colors.textPrimary)
                    
                    Text(episode.title)
                        .font(.subheadline)
                        .foregroundColor(Colors.textSecondary)
                }
                .padding()
                
                Divider()
                
                ScrollView {
                    VStack(spacing: 16) {
                        // Method 1: Watch Ad
                        if !episode.premiumOnly {
                            UnlockMethodCard(
                                icon: "play.rectangle.fill",
                                iconColor: Colors.primary,
                                title: "Watch Ad",
                                subtitle: "FREE - 30 second video",
                                action: {
                                    unlockWithAd()
                                }
                            )
                        }
                        
                        // Method 2: Use Credits
                        if let creditCost = episode.unlockCostCredits, !episode.premiumOnly {
                            UnlockMethodCard(
                                icon: "star.fill",
                                iconColor: .yellow,
                                title: "Use \(creditCost) Credits",
                                subtitle: "Current balance: \(authViewModel.currentUser?.credits ?? 0)",
                                badge: authViewModel.currentUser?.credits ?? 0 < creditCost ? "Insufficient" : nil,
                                action: {
                                    unlockWithCredits()
                                }
                            )
                        }
                        
                        // Method 3: Buy Episode
                        if let iapCost = episode.unlockCostUSD {
                            UnlockMethodCard(
                                icon: "creditcard.fill",
                                iconColor: Colors.success,
                                title: "Buy for $\(String(format: "%.2f", iapCost))",
                                subtitle: "One-time purchase",
                                action: {
                                    unlockWithIAP()
                                }
                            )
                        }
                        
                        // Method 4: Premium
                        if let isPremium = authViewModel.currentUser?.isPremium, !isPremium {
                            UnlockMethodCard(
                                icon: "crown.fill",
                                iconColor: .yellow,
                                title: "Premium Unlimited",
                                subtitle: "$9.99/month - All episodes",
                                action: {
                                    navigateToSubscription()
                                }
                            )
                        }
                        
                        // Error message
                        if let error = errorMessage {
                            Text(error)
                                .font(.caption)
                                .foregroundColor(Colors.accent)
                                .multilineTextAlignment(.center)
                                .padding()
                        }
                    }
                    .padding()
                }
            }
            .background(Colors.background)
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .navigationBarTrailing) {
                    Button("Cancel") {
                        dismiss()
                    }
                    .foregroundColor(Colors.primary)
                }
            }
        }
    }
    
    private func unlockWithAd() {
        isUnlocking = true
        
        AdManager.shared.showRewardedAd { success, adProof in
            if success, let proof = adProof {
                // User watched ad, now call backend to unlock
                Task {
                    await self.callUnlockAPI(method: "ad", adProof: proof)
                }
            } else {
                self.errorMessage = "Ad not available. Try again later."
                self.isUnlocking = false
            }
        }
    }
    
    private func unlockWithCredits() {
        guard let creditCost = episode.unlockCostCredits else { return }
        guard let currentCredits = authViewModel.currentUser?.credits,
              currentCredits >= creditCost else {
            errorMessage = "Insufficient credits. Please top up."
            return
        }
        
        isUnlocking = true
        
        Task {
            await callUnlockAPI(method: "credits")
        }
    }
    
    private func unlockWithIAP() {
        isUnlocking = true
        
        Task {
            do {
                // Get episode unlock product
                let productID = "com.fun.app.episode.unlock"
                guard let product = IAPManager.shared.getProduct(id: productID) else {
                    // If not loaded yet, try to get it directly
                    let products = try await Product.products(for: [productID])
                    guard let product = products.first else {
                        errorMessage = "Product not available"
                        isUnlocking = false
                        return
                    }
                    
                    // Purchase
                    if try await IAPManager.shared.purchase(product) != nil {
                        // Backend verification happens in IAPManager
                        await authViewModel.fetchProfile()
                        dismiss()
                    } else {
                        errorMessage = "Purchase cancelled"
                        isUnlocking = false
                    }
                    return
                }
                
                // Purchase the product
                if try await IAPManager.shared.purchase(product) != nil {
                    // Success - refresh profile and dismiss
                    await authViewModel.fetchProfile()
                    dismiss()
                } else {
                    errorMessage = "Purchase cancelled"
                    isUnlocking = false
                }
            } catch {
                errorMessage = error.localizedDescription
                isUnlocking = false
            }
        }
    }
    
    private func callUnlockAPI(method: String, adProof: String? = nil) async {
        var parameters: [String: Any] = [
            "seriesId": seriesId,
            "episodeNum": episode.episodeNum,
            "method": method
        ]
        
        if let proof = adProof {
            parameters["adProof"] = proof
        }
        
        do {
            let response: UnlockResponse = try await APIClient.shared.request(
                .unlock,
                method: .post,
                parameters: parameters
            )
            
            // Success! Refresh user profile and dismiss
            await authViewModel.fetchProfile()
            isUnlocking = false
            dismiss()
        } catch {
            errorMessage = error.localizedDescription
            isUnlocking = false
        }
    }
    
    private func navigateToSubscription() {
        // Dismiss current sheet and show subscription
        dismiss()
        
        // Post notification to show subscription view
        NotificationCenter.default.post(name: NSNotification.Name("ShowSubscription"), object: nil)
    }
}

struct UnlockMethodCard: View {
    let icon: String
    let iconColor: Color
    let title: String
    let subtitle: String
    var badge: String? = nil
    let action: () -> Void
    
    var body: some View {
        Button(action: action) {
            HStack(spacing: 16) {
                Image(systemName: icon)
                    .font(.title2)
                    .foregroundColor(iconColor)
                    .frame(width: 50, height: 50)
                    .background(iconColor.opacity(0.2))
                    .cornerRadius(12)
                
                VStack(alignment: .leading, spacing: 4) {
                    HStack {
                        Text(title)
                            .font(.headline)
                            .foregroundColor(Colors.textPrimary)
                        
                        if let badge = badge {
                            Text(badge)
                                .font(.caption.weight(.semibold))
                                .foregroundColor(Colors.accent)
                                .padding(.horizontal, 8)
                                .padding(.vertical, 4)
                                .background(Colors.accent.opacity(0.2))
                                .cornerRadius(4)
                        }
                    }
                    
                    Text(subtitle)
                        .font(.subheadline)
                        .foregroundColor(Colors.textSecondary)
                }
                
                Spacer()
                
                Image(systemName: "chevron.right")
                    .font(.caption.weight(.semibold))
                    .foregroundColor(Colors.textSecondary)
            }
            .padding()
            .background(Colors.surface)
            .cornerRadius(16)
        }
        .disabled(badge != nil)
    }
}

File: mobile/android/FUN/app/src/main/java/com/fun/app/ui/components/player/UnlockSheet.kt
/**
 * FUN App - Unlock Sheet (Android)
 * Modal bottom sheet with 4 unlock methods
 */

package com.fun.app.ui.components.player

import android.app.Activity
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.unit.dp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import com.fun.app.data.models.Episode
import com.fun.app.ui.theme.*
import com.fun.app.FunApplication

@Composable
fun UnlockSheet(
    episode: Episode,
    seriesId: String,
    onDismiss: () -> Unit
) {
    val context = androidx.compose.ui.platform.LocalContext.current
    val activity = context as? Activity
    val unlockViewModel = remember { com.fun.app.ui.viewmodels.UnlockViewModel() }
    
    val isUnlocking by unlockViewModel.isUnlocking.collectAsState()
    val errorMessage by unlockViewModel.errorMessage.collectAsState()
    val unlockSuccess by unlockViewModel.unlockSuccess.collectAsState()
    
    // Close sheet on success
    LaunchedEffect(unlockSuccess) {
        if (unlockSuccess) {
            onDismiss()
        }
    }
    
    DisposableEffect(Unit) {
        onDispose {
            unlockViewModel.resetState()
        }
    }
    
    Dialog(
        onDismissRequest = onDismiss,
        properties = DialogProperties(usePlatformDefaultWidth = false)
    ) {
        Surface(
            modifier = Modifier
                .fillMaxWidth()
                .fillMaxHeight(0.75f),
            shape = RoundedCornerShape(topStart = 24.dp, topEnd = 24.dp),
            color = MaterialTheme.colorScheme.background
        ) {
            Column(
                modifier = Modifier.fillMaxSize()
            ) {
                // Header
                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(24.dp),
                    horizontalAlignment = Alignment.CenterHorizontally,
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    // Drag handle
                    Surface(
                        modifier = Modifier
                            .width(40.dp)
                            .height(4.dp),
                        shape = RoundedCornerShape(2.dp),
                        color = TextSecondary.copy(alpha = 0.3f)
                    ) {}
                    
                    Spacer(modifier = Modifier.height(8.dp))
                    
                    Text(
                        text = "Unlock Episode ${episode.episodeNum}",
                        style = MaterialTheme.typography.headlineSmall,
                        color = TextPrimary
                    )
                    
                    Text(
                        text = episode.title,
                        style = MaterialTheme.typography.bodyMedium,
                        color = TextSecondary
                    )
                }
                
                Divider(color = Divider)
                
                // Unlock methods
                Column(
                    modifier = Modifier
                        .weight(1f)
                        .verticalScroll(rememberScrollState())
                        .padding(16.dp),
                    verticalArrangement = Arrangement.spacedBy(16.dp)
                ) {
                    // Method 1: Watch Ad
                    if (!episode.premiumOnly) {
                        UnlockMethodCard(
                            icon = Icons.Default.PlayArrow,
                            iconColor = Primary,
                            iconBackground = Primary.copy(alpha = 0.2f),
                            title = "Watch Ad",
                            subtitle = "FREE - 30 second video",
                            onClick = {
                                activity?.let { act ->
                                    FunApplication.instance.adManager.showRewardedAd(act) { success, adProof ->
                                        if (success && adProof != null) {
                                            unlockViewModel.unlockWithAd(seriesId, episode.episodeNum, adProof)
                                        }
                                    }
                                }
                            }
                        )
                    }
                    
                    // Method 2: Use Credits
                    episode.unlockCostCredits?.let { cost ->
                        if (!episode.premiumOnly) {
                            UnlockMethodCard(
                                icon = Icons.Default.Star,
                                iconColor = Color.Yellow,
                                iconBackground = Color.Yellow.copy(alpha = 0.2f),
                                title = "Use $cost Credits",
                                subtitle = "Current balance: 150", // TODO: Get from user state
                                onClick = {
                                    unlockViewModel.unlockWithCredits(seriesId, episode.episodeNum)
                                }
                            )
                        }
                    }
                    
                    // Method 3: Buy Episode
                    episode.unlockCostUSD?.let { cost ->
                        UnlockMethodCard(
                            icon = Icons.Default.ShoppingCart,
                            iconColor = Success,
                            iconBackground = Success.copy(alpha = 0.2f),
                            title = "Buy for $${String.format("%.2f", cost)}",
                            subtitle = "One-time purchase",
                            onClick = {
                                // TODO: IAP episode purchase
                            }
                        )
                    }
                    
                    // Method 4: Premium
                    UnlockMethodCard(
                        icon = Icons.Default.Star,
                        iconColor = Color.Yellow,
                        iconBackground = Color.Yellow.copy(alpha = 0.2f),
                        title = "Premium Unlimited",
                        subtitle = "$9.99/month - All episodes",
                        onClick = {
                            onDismiss()
                            // TODO: Navigate to subscription
                        }
                    )
                    
                    // Error message
                    errorMessage?.let { error ->
                        Text(
                            text = error,
                            style = MaterialTheme.typography.bodySmall,
                            color = Accent,
                            modifier = Modifier.padding(16.dp)
                        )
                    }
                }
                
                // Close button
                TextButton(
                    onClick = onDismiss,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(16.dp)
                ) {
                    Text("Cancel")
                }
            }
        }
    }
}

@Composable
private fun UnlockMethodCard(
    icon: ImageVector,
    iconColor: Color,
    iconBackground: Color,
    title: String,
    subtitle: String,
    badge: String? = null,
    onClick: () -> Unit
) {
    Card(
        onClick = onClick,
        modifier = Modifier.fillMaxWidth(),
        colors = CardDefaults.cardColors(
            containerColor = Surface
        )
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(16.dp),
            horizontalArrangement = Arrangement.spacedBy(16.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Surface(
                modifier = Modifier.size(50.dp),
                shape = RoundedCornerShape(12.dp),
                color = iconBackground
            ) {
                Icon(
                    imageVector = icon,
                    contentDescription = null,
                    tint = iconColor,
                    modifier = Modifier
                        .fillMaxSize()
                        .padding(12.dp)
                )
            }
            
            Column(
                modifier = Modifier.weight(1f),
                verticalArrangement = Arrangement.spacedBy(4.dp)
            ) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    Text(
                        text = title,
                        style = MaterialTheme.typography.titleMedium,
                        color = TextPrimary
                    )
                    
                    badge?.let {
                        Surface(
                            shape = RoundedCornerShape(4.dp),
                            color = Accent.copy(alpha = 0.2f)
                        ) {
                            Text(
                                text = it,
                                style = MaterialTheme.typography.labelSmall,
                                color = Accent,
                                modifier = Modifier.padding(horizontal = 8.dp, vertical = 4.dp)
                            )
                        }
                    }
                }
                
                Text(
                    text = subtitle,
                    style = MaterialTheme.typography.bodyMedium,
                    color = TextSecondary
                )
            }
            
            Icon(
                imageVector = Icons.Default.KeyboardArrowRight,
                contentDescription = null,
                tint = TextSecondary
            )
        }
    }
}
